doc/man/bitcoind.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.8.
.TH BITCOIND "1" "February 2024" "bitcoin-tx v2.1.0.0" "User Commands"
.SH NAME
bitcoind \- manual page for bitcoin-tx v2.1.0.0 
.SH DESCRIPTION
TBCNODE version v2.1.0.0\-dirty\-dev
.SS "Usage:"
.TP
bitcoind [options]
Start TBCNODE Daemon
.SH OPTIONS
.HP
\-?
.IP
Print this help message and exit
.HP
\fB\-version\fR
.IP
Print version and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when a relevant alert is received or we see a really
long fork (%s in cmd is replaced by message)
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-assumevalid=\fR<hex>
.IP
If this block is in the chain assume that it and its ancestors are valid
and potentially skip their script verification (0 to verify all,
default:
000000000000000000e45ad2fbcc5ff3e85f0868dd8f00ad4e92dffabe28f8d2,
testnet:
0000000000327972b8470c11755adf8f4319796bafae01f5a6650490b98a17db)
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: bitcoin.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-dbcache=\fR<n>
.IP
Set database cache size in megabytes (4 to 16384, default: 450). The
value may be given in megabytes or with unit (B, KiB, MiB, GiB).
.HP
\fB\-genesisactivationheight\fR
.IP
Set block height at which genesis should be activated. (default:
620538).
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external blk000??.dat file on startup
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the transaction memory pool below <n> megabytes (default: 1000,
must be at least 300). The value may be given in megabytes or
with unit (B, kB, MB, GB).
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
336)
.HP
\fB\-maxmempoolnonfinal=\fR<n>
.IP
Keep the non\-final transaction memory pool below <n> megabytes (default:
50). The value may be given in megabytes or with unit (B, KiB,
MiB, GiB).
.HP
\fB\-mempoolexpirynonfinal=\fR<n>
.IP
Do not keep transactions in the non\-final mempool longer than <n> hours
(default: 672)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-threadsperblock=\fR<n>
.IP
Set the number of script verification threads used when validating
single block (0 to 64, 0 = auto, default: 0)
.HP
\fB\-scriptvalidatormaxbatchsize=\fR<n>
.IP
Set size of script verification batch per thread (1 to 255, default:
128)
.HP
\fB\-maxparallelblocks=\fR<n>
.IP
Set the number of block that can be validated in parallel across all
nodes. If additional block arrive, validation of an old block is
terminated. (1 to 100, default: 4)
.HP
\fB\-maxparallelblocksperpeer=\fR<n>
.IP
Set the number of blocks that can be validated in parallel from a single
peer. If peers sends another block, the validation of it is
delayed. (1 to maxparallelblocks, default: 3)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file (default: bitcoind.pid)
.HP
\fB\-preload=\fR<n>
.IP
If n is set to 1, blockchain state will be preloaded into memory. If n
is 0, no preload will happen. Other values for n are not allowed.
The default value is 0. This option is not supported on Windows
operating systems.
.HP
\fB\-prune=\fR<n>
.IP
Reduce storage requirements by enabling pruning (deleting) of old
blocks. This allows the pruneblockchain RPC to be called to
delete specific blocks, and enables automatic pruning of old
blocks if a target size in MiB is provided. This mode is
incompatible with \fB\-txindex\fR and \fB\-rescan\fR. Warning: Reverting this
setting requires re\-downloading the entire blockchain. (default:
0 = disable pruning blocks, 1 = allow manual pruning via RPC,
>550 = automatically prune block files to stay under the
specified target size in MiB)
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks
.HP
\fB\-reindex\fR
.IP
Rebuild chain state and block index from the blk*.dat files on disk
.HP
\fB\-rejectmempoolrequest\fR
.IP
Reject every mempool request from non\-whitelisted peers.
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 0)
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open
.HP
\fB\-banscore=\fR<n>
.IP
Threshold for disconnecting misbehaving peers (default: 100)
.HP
\fB\-bantime=\fR<n>
.IP
Number of seconds to keep misbehaving peers from reconnecting (default:
86400)
.HP
\fB\-bind=\fR<addr>
.IP
Bind to given address and always listen on it. Use [host]:port notation
for IPv6
.HP
\fB\-blockstallingmindownloadspeed=\fR<n>
.IP
Minimum average download speed (Kbytes/s) we will allow a stalling peer
to fall to during IBD. A value of 0 means stall detection is
disabled (default: 100Kb/s)
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node(s); \fB\-noconnect\fR or \fB\-connect\fR=\fI\,0\/\fR alone to
disable automatic connections
.HP
\fB\-discover\fR
.IP
Discover own IP addresses (default: 1 when listening and no \fB\-externalip\fR
or \fB\-proxy\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR/\-noconnect)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 1)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or
\fB\-connect\fR/\-noconnect)
.HP
\fB\-listenonion\fR
.IP
Automatically create Tor hidden service (default: 1)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (default: 125)
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer in kilobytes (default: 5000). The
value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer in kilobytes (default: 1000). The
value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-factormaxsendqueuesbytes=\fR<n>
.IP
Factor that will be multiplied with excessiveBlockSize to limit the
maximum bytes in all sending queues. If this size is exceeded, no
response to block related P2P messages is sent. (default factor:
4)
.HP
\fB\-maxtimeadjustment\fR
.IP
Maximum allowed median peer time offset adjustment. Local perspective of
time may be influenced by peers forward or backward by this
amount. (default: 4200 seconds)
.HP
\fB\-broadcastdelay=\fR<n>
.IP
Set inventory broadcast delay duration in millisecond(min: 0, max:
50000)
.HP
\fB\-onion=\fR<ip:port>
.IP
Use separate SOCKS5 proxy to reach peers via Tor hidden services
(default: \fB\-proxy\fR)
.HP
\fB\-onlynet=\fR<net>
.IP
Only connect to nodes in network <net> (ipv4, ipv6 or onion)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
1)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port> (default: 8333 or testnet: 18333)
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. This enables Tor
stream isolation (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect
.HP
\fB\-timeout=\fR<n>
.IP
Specify connection timeout in milliseconds (minimum: 1, default: 5000)
.HP
\fB\-torcontrol=\fR<ip>:<port>
.IP
Tor control port to use if onion listening enabled (default:
127.0.0.1:9051)
.HP
\fB\-torpassword=\fR<pass>
.IP
Tor control port password (default: empty)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 0)
.HP
\fB\-whitebind=\fR<addr>
.IP
Bind to given address and whitelist peers connecting to it. Use
[host]:port notation for IPv6
.HP
\fB\-whitelist=\fR<IP address or network>
.IP
Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
CIDR notated network (e.g. 1.2.3.0/24). Can be specified multiple
times. Whitelisted peers cannot be DoS banned and their
transactions are always relayed, even if they are already in the
mempool, useful e.g. for a gateway
.HP
\fB\-whitelistrelay\fR
.IP
Accept relayed transactions received from whitelisted peers even when
not relaying transactions (default: 1)
.HP
\fB\-whitelistforcerelay\fR
.IP
Force relay of transactions from whitelisted peers even if they violate
local relay policy (default: 1)
.HP
\fB\-maxuploadtarget=\fR<n>
.IP
Tries to keep outbound traffic under the given target (in MiB per 24h),
0 = no limit (default: 0). The value may be given in megabytes or
with unit (KiB, MiB, GiB).
.PP
Wallet options:
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 1000)
.HP
\fB\-fallbackfee=\fR<amt>
.IP
A fee rate (in BSV/kB) that will be used when fee estimation has
insufficient data (default: 0.0002)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fees (in BSV/kB) smaller than this are considered zero fee for
transaction creation (default: 0.00001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee (in BSV/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-rescan\fR
.IP
Rescan the block chain for missing wallet transactions on startup
.HP
\fB\-salvagewallet\fR
.IP
Attempt to recover private keys from a corrupt wallet on startup
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-usehd\fR
.IP
Use hierarchical deterministic key generation (HD) after BIP32. Only has
effect during wallet creation/first start (default: 1)
.HP
\fB\-upgradewallet\fR
.IP
Upgrade wallet to latest format on startup
.HP
\fB\-wallet=\fR<file>
.IP
Specify wallet file (within data directory) (default: wallet.dat)
.HP
\fB\-walletbroadcast\fR
.IP
Make the wallet broadcast transactions (default: 1)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes (%s in cmd is replaced
by TxID)
.HP
\fB\-zapwallettxes=\fR<mode>
.IP
Delete all wallet transactions and only recover those parts of the
blockchain through \fB\-rescan\fR on startup (1 = keep tx meta data e.g.
account owner and payment request information, 2 = drop tx meta
data)
.PP
ZeroMQ notification options:
.HP
\fB\-zmqpubhashblock=\fR<address>
.IP
Enable publish hash block in <address>
.HP
\fB\-zmqpubhashtx=\fR<address>
.IP
Enable publish hash transaction in <address>
.HP
\fB\-zmqpubrawblock=\fR<address>
.IP
Enable publish raw block in <address>
.HP
\fB\-zmqpubrawtx=\fR<address>
.IP
Enable publish raw transaction in <address>
.PP
Debugging/Testing options:
.HP
\fB\-uacomment=\fR<cmt>
.IP
Append comment to the user agent string
.HP
\fB\-debug=\fR<category>
.IP
Output debugging information (default: 0, supplying <category> is
optional). If <category> is not supplied or if <category> = 1,
output all debugging information.<category> can be: net, tor,
mempool, http, bench, zmq, db, rpc, addrman, selectcoins,
reindex, cmpctblock, rand, prune, proxy, mempoolrej, libevent,
coindb, leveldb, txnprop, txnsrc, journal, txnval.
.HP
\fB\-debugexclude=\fR<category>
.IP
Exclude debugging information for a category. Can be used in conjunction
with \fB\-debug\fR=\fI\,1\/\fR to output debug logs for all categories except one
or more specified categories.
.HP
\fB\-help\-debug\fR
.IP
Show all debugging options (usage: \fB\-\-help\fR \fB\-help\-debug\fR)
.HP
\fB\-debugp2pthreadstalls\fR
.IP
Log P2P requests that stall request processing loop for longer than
specified milliseconds (default: disabled)
.HP
\fB\-logips\fR
.IP
Include IP addresses in debug output (default: 0)
.HP
\fB\-logtimestamps\fR
.IP
Prepend debug output with timestamp (default: 1)
.HP
\fB\-minrelaytxfee=\fR<amt>
.IP
Fees (in BSV/kB) smaller than this are considered zero fee for relaying,
mining and transaction creation (default: 0.0000025)
.HP
\fB\-maxtxfee=\fR<amt>
.IP
Maximum total fees (in BSV) to use in a single wallet transaction or raw
transaction; setting this too low may abort large transactions
(default: 0.10)
.HP
\fB\-printtoconsole\fR
.IP
Send trace/debug info to console instead of bitcoind.log file
.HP
\fB\-shrinkdebugfile\fR
.IP
Shrink bitcoind.log file on client startup (default: 1 when no \fB\-debug\fR)
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.HP
\fB\-regtest\fR
.IP
Enter regression test mode, which uses a special chain in which blocks
can be solved instantly. This is intended for regression testing
tools and app development.
.HP
\fB\-stn\fR
.IP
Use the Scaling Test Network
.PP
Node relay options:
.HP
\fB\-excessiveblocksize=\fR<n>
.IP
Set the maximum block size in bytes we will accept from any source. This
is the effective block size hard limit and it is a required
parameter (0 = unlimited). The value may be given in bytes or
with unit (B, kB, MB, GB).
.HP
\fB\-datacarrier\fR
.IP
Relay and mine data carrier transactions (default: 1)
.HP
\fB\-datacarriersize\fR
.IP
Maximum size of data in data carrier transactions we relay and mine
(default: 4294967295). The value may be given in bytes or with
unit (B, kB, MB, GB).
.HP
\fB\-maxstackmemoryusageconsensus\fR
.IP
Set maximum stack memory usage in bytes used for script verification
we're willing to accept from any source (0 = unlimited) after
Genesis is activated (consensus level). This is a required
parameter. The value may be given in bytes or with unit (B, kB,
MB, GB).
.HP
\fB\-maxstackmemoryusagepolicy\fR
.IP
Set maximum stack memory usage used for script verification we're
willing to relay/mine in a single transaction (default: 100 MB, 0
= unlimited) after Genesis is activated (policy level). The value
may be given in bytes or with unit (B, kB, MB, GB).Must be less
or equal to \fB\-maxstackmemoryusageconsensus\fR.
.HP
\fB\-maxopsperscriptpolicy=\fR<n>
.IP
Set maximum number of non\-push operations we're willing to relay/mine
per script (default: unlimited, 0 = unlimited), after Genesis is
activated
.HP
\fB\-maxtxsigopscountspolicy=\fR<n>
.IP
Set maximum allowed number of signature operations we're willing to
relay/mine in a single transaction (default: unlimited, 0 =
unlimited) after Genesis is activated.
.HP
\fB\-maxstdtxvalidationduration=\fR<n>
.IP
Set the single standard transaction validation duration threshold in
milliseconds after which the standard transaction validation will
terminate with error and the transaction is not accepted to
mempool (min 5ms, default: 10ms)
.HP
\fB\-maxnonstdtxvalidationduration=\fR<n>
.IP
Set the single non\-standard transaction validation duration threshold in
milliseconds after which the non\-standard transaction validation
will terminate with error and the transaction is not accepted to
mempool (min 10ms, default: 1000ms)
.HP
\fB\-maxtxsizepolicy=\fR<n>
.IP
Set maximum transaction size in bytes we relay and mine (default: 10 MB,
min: 99999 B, 0 = unlimited) after Genesis is activated. The
value may be given in bytes or with unit (B, kB, MB, GB).
.HP
\fB\-minconsolidationfactor=\fR<n>
.IP
Set minimum ratio between sum of utxo scriptPubKey sizes spent in a
consolidation transaction, to the corresponding sum of output
scriptPubKey sizes. The ratio between number of consolidation
transaction inputs to the number of outputs also needs to be
greater or equal to the minimum consolidation factor (default:
20). A value of 0 disables free consolidation transactions
.HP
\fB\-maxconsolidationinputscriptsize=\fR<n>
.IP
This number is the maximum length for a scriptSig input in a
consolidation txn (default: 150).
.HP
\fB\-minconsolidationinputmaturity=\fR<n>
.IP
Minimum number of confirmations of inputs spent by consolidation
transactions (default: 6).
.HP
\fB\-acceptnonstdconsolidationinput=\fR<n>
.IP
Accept consolidation transactions spending non standard inputs (default:
0).
.HP
\fB\-maxscriptsizepolicy\fR
.IP
Set maximum script size in bytes we're willing to relay/mine per script
after Genesis is activated. (default: 10000, 0 = unlimited). The
value may be given in bytes or with unit (B, kB, MB, GB).
.HP
\fB\-maxscriptnumlengthpolicy=\fR<n>
.IP
Set maximum allowed number length we're willing to relay/mine in scripts
(default: 250000, 0 = unlimited) after Genesis is activated. The
value may be given in bytes or with unit (B, kB, MB, GB).
.PP
Block creation options:
.HP
\fB\-blockmaxsize=\fR<n>
.IP
Set maximum block size in bytes we will mine. Size of the mined block
will never exceed the maximum block size we will accept
(\fB\-excessiveblocksize\fR). The value may be given in bytes or with
unit (B, kB, MB, GB). If not specified, the following defaults
are used: Mainnet: 32 MB before 2019\-07\-24 14:00:00 and 128 MB
after, Testnet: 32 MB before 2019\-07\-24 14:00:00 and 128 MB
after.
.HP
\fB\-blockprioritypercentage=\fR<n>
.IP
Set maximum percentage of a block reserved to high\-priority/low\-fee
transactions (default: 5). NOTE: This is supported only by the
legacy block assembler which is not default block assembler any
more and will be removed in the upcoming release.
.HP
\fB\-blockmintxfee=\fR<amt>
.IP
Set lowest fee rate (in BSV/kB) for transactions to be included in block
creation. (default: 0.000005)
.HP
\fB\-invalidateblock=\fR<hash>
.IP
Permanently marks an existing block as invalid as if it violated a
consensus rule (same as InvalidateBlock RPC function). If
specified block header was not received yet, the header will be
ignored when it is received from a peer. This option can be
specified multiple times.
.HP
\fB\-banclientua=\fR<ua>
.IP
Ban clients whose User Agent contains specified string (case
insensitive). This option can be specified multiple times.
.HP
\fB\-blockassembler=\fR<type>
.IP
Set the type of block assembler to use for mining. Supported options are
LEGACY or JOURNALING. (default: JOURNALING)
.HP
\fB\-jbamaxtxnbatch=\fR<max batch size>
.IP
Set the maximum number of transactions processed in a batch by the
journaling block assembler (default: 20000)
.PP
RPC server options:
.HP
\fB\-server\fR
.IP
Accept command line and JSON\-RPC commands
.HP
\fB\-rest\fR
.IP
Accept public REST requests (default: 0)
.HP
\fB\-rpcbind=\fR<addr>
.IP
Bind to given address to listen for JSON\-RPC connections. Use
[host]:port notation for IPv6. This option can be specified
multiple times (default: bind to all interfaces)
.HP
\fB\-rpccookiefile=\fR<loc>
.IP
Location of the auth cookie (default: data dir)
.HP
\fB\-rpcuser=\fR<user>
.IP
Username for JSON\-RPC connections
.HP
\fB\-rpcpassword=\fR<pw>
.IP
Password for JSON\-RPC connections
.HP
\fB\-rpcauth=\fR<userpw>
.IP
Username and hashed password for JSON\-RPC connections. The field
<userpw> comes in the format: <USERNAME>:<SALT>$<HASH>. A
canonical python script is included in share/rpcuser. The client
then connects normally using the
rpcuser=<USERNAME>/rpcpassword=<PASSWORD> pair of arguments. This
option can be specified multiple times
.HP
\fB\-rpcport=\fR<port>
.IP
Listen for JSON\-RPC connections on <port> (default: 8332 or testnet:
18332)
.HP
\fB\-rpcallowip=\fR<ip>
.IP
Allow JSON\-RPC connections from specified source. Valid for <ip> are a
single IP (e.g. 1.2.3.4), a network/netmask (e.g.
1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This
option can be specified multiple times
.HP
\fB\-magicbytes=\fR<hexcode>
.IP
Allow users to split the test net by changing the magicbytes. This
option only work on a network different than mainnet. default :
0f0f0f0f
.HP
\fB\-rpcthreads=\fR<n>
.IP
Set the number of threads to service RPC calls (default: 4)
.HP
\fB\-rpccorsdomain\fR=\fI\,value\/\fR
.IP
Domain from which to accept cross origin requests (browser enforced)
.HP
\fB\-invalidcsinterval=\fR<n>
.IP
Set the time limit on the reception of invalid message checksums from a
single node in milliseconds (default: 500ms)
.HP
\fB\-invalidcsfreq=\fR<n>
.IP
Set the limit on the number of invalid checksums received over a given
time period from a single node  (default: 100)
.PP
Orphan txns config :
.HP
\fB\-blockreconstructionextratxn=\fR<n>
.IP
Extra transactions to keep in memory for compact block reconstructions
(default: 100)
.HP
\fB\-maxorphantxsize=\fR<n>
.IP
Keep at most <n> MB of unconnectable transactions in memory (default:
100 MB). The value may be given in megabytes or with unit (B, kB,
MB, GB).
.HP
\fB\-maxcollectedoutpoints=\fR<n>
.IP
Keep at most <n> collected outpoints in memory (default: 300000)
.PP
TxnValidator options:
.HP
\fB\-numstdtxvalidationthreads=\fR<n>
.IP
Set the number of 'High' priority threads used to validate standard txns
(dynamically calculated default: 9)
.HP
\fB\-numnonstdtxvalidationthreads=\fR<n>
.IP
Set the number of 'Low' priority threads used to validate non\-standard
txns (dynamically calculated default: 3)
.HP
\fB\-maxstdtxnsperthreadratio=\fR<n>
.IP
Set the max ratio for a number of standard txns per 'High' priority
thread (default: 1000)
.HP
\fB\-maxnonstdtxnsperthreadratio=\fR<n>
.IP
Set the max ratio for a number of non\-standard txns per 'Low' priority
thread (default: 1000)
.HP
\fB\-txnvalidationasynchrunfreq=\fR<n>
.IP

doc/release-notes.md
# TBCNODE software version 2.0.0 Release Notes

## List of Changes
* Many, see git diff

## Scaling Test Network (STN) Reset
N/A

# Previous Releases
* [Version 0.1.0](release-notes-v0.1.0.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/0.1.0/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v0.1.0) - 2018-10-15
* [Version 0.1.1](release-notes-v0.1.1.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/0.1.1/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v0.1.1) - 2019-02-11
* [Version 0.2.0](release-notes-v0.2.0.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/0.2.0/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v0.2.0) - 2019-06-05
* [Version 0.2.1](release-notes-v0.2.1.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/0.2.1/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v0.2.1) - 2019-07-12
* [Version 0.2.2.beta](release-notes-v0.2.2-beta.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/0.2.2.beta/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v0.2.2.beta) - 2019-10-30
* [Version 1.0.0](release-notes-v1.0.0.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/1.0.0/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v1.0.0) - 2020-01-15
* [Version 1.0.1](release-notes-v1.0.1.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/1.0.1/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v1.0.1) - 2020-01-28
* [Version 1.0.2](release-notes-v1.0.2.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/1.0.2/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v1.0.2) - 2020-02-17
* [Version 1.0.3](release-notes-v1.0.3.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/1.0.3/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v1.0.3) - 2020-04-28
* [Version 1.0.4](release-notes-v1.0.4.md) - [Download](https://download.bitcoinsv.io/bitcoinsv/1.0.4/) - [Source](https://github.com/bitcoin-sv/bitcoin-sv/tree/v1.0.4) - 2020-07-01
* [Version 1.0.5RC](release-notes-v1.0.5RC.md) - [Source](https://github.com/TuringBitChain/TBCNODE/tree/598ad0124854abe9a8293735f51593d83b15a292)  
* [Version 1.9.9](release-notes.md) - [Source](https://github.com/TuringBitChain/TBCNODE/tree/1f32b20c446f31ac07f12790e3ef7b3b27b3cfa6)  
* [Version 2.0.0](release-notes.md) - [Source](https://github.com/TuringBitChain/TBCNODE/tree/a28b623d95852fa7d6246f41f74c990eebb4f4b8)  


src/bitcoind.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2016 The Bitcoin Core developers
// Copyright (c) 2019 Bitcoin Association
// Distributed under the Open TBC software license, see the accompanying LICENSE.

#if defined(HAVE_CONFIG_H)
#include "config/bitcoin-config.h"
#endif

#include "chainparams.h"
#include "clientversion.h"
#include "compat.h"
#include "config.h"
#include "fs.h"
#include "httprpc.h"
#include "httpserver.h"
#include "init.h"
#include "noui.h"
#include "rpc/server.h"
#include "scheduler.h"
#include "util.h"
#include "utilstrencodings.h"

#include <boost/algorithm/string/predicate.hpp>
#include <boost/thread.hpp>

#include <cstdio>


/* Introduction text for doxygen: */

/*! \mainpage Developer documentation
 *
 * \section intro_sec Introduction
 *
 * This is the developer documentation of Bitcoin SV
 * (https://bitcoinsv.io/). Bitcoin SV is a client for the digital
 * currency called Bitcoin SV, which enables
 * instant payments to anyone, anywhere in the world. Bitcoin SV uses
 * peer-to-peer technology to operate with no central authority: managing
 * transactions and issuing money are carried out collectively by the network.
 *
 * The software is a community-driven open source project, released under the
 * Open BSV license.
 *
 * \section Navigation
 * Use the buttons <code>Namespaces</code>, <code>Classes</code> or
 * <code>Files</code> at the top of the page to start navigating the code.
 */

void WaitForShutdown(boost::thread_group *threadGroup, const task::CCancellationToken& shutdownToken) {

    // Tell the main threads to shutdown.
    while (!shutdownToken.IsCanceled()) {
        MilliSleep(200);
    }
    if (threadGroup) {
        Interrupt(*threadGroup);
        threadGroup->join_all();
    }
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//
bool AppInit(int argc, char *argv[]) {
    RenameThread("tbc-main");
    boost::thread_group threadGroup;
    CScheduler scheduler;

    // FIXME: Ideally, we'd like to build the config here, but that's currently
    // not possible as the whole application has too many global state. However,
    // this is a first step.
    auto& config = GlobalConfig::GetConfig();

    bool fRet = false;

    //
    // Parameters
    //
    // main()
    gArgs.ParseParameters(argc, argv);

    // Process help and version before taking care about datadir
    if (gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") ||
        gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version")) {
        std::string strUsage = strprintf(_("%s"), _(PACKAGE_NAME)) +
                               " " + _("version") + " " + FormatFullVersion() +
                               "\n";

        if (gArgs.IsArgSet("-version")) {
            strUsage += FormatParagraph(LicenseInfo());
        } else {
            strUsage += "\n" + _("Usage:") + "\n" +
                        "  bitcoind [options]                     " +
                        strprintf(_("Start %s Daemon"), _(PACKAGE_NAME)) + "\n";

            strUsage += "\n" + HelpMessage(HMM_BITCOIND);
        }

        fprintf(stdout, "%s", strUsage.c_str());
        return true;
    }

    try {
        if (!fs::is_directory(GetDataDir(false))) {
            fprintf(stderr,
                    "Error: Specified data directory \"%s\" does not exist.\n",
                    gArgs.GetArg("-datadir", "").c_str());
            return false;
        }
        try {
            gArgs.ReadConfigFile(gArgs.GetArg("-conf", BITCOIN_CONF_FILENAME));
        } catch (const std::exception &e) {
            fprintf(stderr, "Error reading configuration file: %s\n", e.what());
            return false;
        }
        // Check for -testnet or -regtest parameter (Params() calls are only
        // valid after this clause)
        try {
            SelectParams(ChainNameFromCommandLine());
        } catch (const std::exception &e) {
            fprintf(stderr, "Error: %s\n", e.what());
            return false;
        }

        // Fill config with block size data
        config.SetDefaultBlockSizeParams(Params().GetDefaultBlockSizeParams());

        // maxstackmemoryusageconsensus and excessiveblocksize are required parameters
        if (!gArgs.IsArgSet("-maxstackmemoryusageconsensus") || !gArgs.IsArgSet("-excessiveblocksize"))
        {
            fprintf(stderr, "Mandatory consensus parameter is not set. In order to start bitcoind you must set the "
                            "following consensus parameters: \"excessiveblocksize\" and "
                            "\"maxstackmemoryusageconsensus\". In order to start bitcoind with no limits you can set "
                            "both of these parameters to 0 however it is strongly recommended to ensure you understand "
                            "the implications of this setting.\n\n"
                            "For more information of how to choose these settings safely for your use case refer to: "
                            "https://bitcoinsv.io/choosing-consensus-settings/");
            return false;
        }

        // Command-line RPC
        bool fCommandLine = false;
        for (int i = 1; i < argc; i++)
            if (!IsSwitchChar(argv[i][0]) &&
                !boost::algorithm::istarts_with(argv[i], "bitcoin:"))
                fCommandLine = true;

        if (fCommandLine) {
            fprintf(stderr, "Error: There is no RPC client functionality in "
                            "bitcoind anymore. Use the bitcoin-cli utility "
                            "instead.\n");
            exit(EXIT_FAILURE);
        }
        // -server defaults to true for bitcoind
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitParameterInteraction(config)) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitSanityChecks()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (gArgs.GetBoolArg("-daemon", false)) {
#if HAVE_DECL_DAEMON
            fprintf(stdout, "TuringBitChain starting\n");

            // Daemonize
            if (daemon(1, 0)) {
                // don't chdir (1), do close FDs (0)
                fprintf(stderr, "Error: daemon() failed: %s\n",
                        strerror(errno));
                return false;
            }
#else
            fprintf(
                stderr,
                "Error: -daemon is not supported on this operating system\n");
            return false;
#endif // HAVE_DECL_DAEMON
        }

        fRet = AppInitMain(config, threadGroup, scheduler, GetShutdownToken());
    } catch (const std::exception &e) {
        PrintExceptionContinue(&e, "AppInit()");
    } catch (...) {
        PrintExceptionContinue(nullptr, "AppInit()");
    }

    if (!fRet) {
        Interrupt(threadGroup);
        // threadGroup.join_all(); was left out intentionally here, because we
        // didn't re-test all of the startup-failure cases to make sure they
        // don't result in a hang due to some
        // thread-blocking-waiting-for-another-thread-during-startup case.
    } else {
        LogPrintf("Preload wait for shutdown\n");
        WaitForShutdown(&threadGroup, GetShutdownToken());
        LogPrintf("Preload wait for shutdown done\n");
    }
    LogPrintf("Checking Thread shutdown\n");
    Shutdown();

    return fRet;
}

int main(int argc, char *argv[]) {
    SetupEnvironment();

    // Connect bitcoind signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
}
